### PR TITLE
WIP [#769]: comment out SPNameQualifier from SimpleSAML lib for testing

### LINKED ISSUE #769: Expose SPNameQualifier setting to the SAML settings
Facing an issue when using the Oracle Identity Manager as the IdP. The user cannot log out of Moodle.

When debugging, we have come across the following response: Provider not part of the Affiliation

After some more testing/debugging in Moodle, we found that when sending the Logout request to the IdP, the SPNameQualifier and NameQualifier attributes are sent with the request. When we hacked the values deep in the SimpleSAML library to not send those attributes. This proved successful in logging the user out correctly.

Example logout request captured:

Unsuccessful
```
<samlp:LogoutRequest xmlns:samlp="urn:oasis:names:tc:SAML:2.0:protocol"
                   xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion"
                   ID="_40bf2ce71cf77b761ce21ce025e51edc4eb6f6f29a"
                   Version="2.0"
                   IssueInstant="2023-08-22T04:05:01Z"
                   Destination="https://IdPdomain" >
   <saml:Issuer>https://moodledomain/auth/saml2/sp/metadata.php</saml:Issuer>
   <saml:NameID NameQualifier="https://moodledomain/auth/saml2/sp/metadata.php"
             SPNameQualifier="https://IdPdomain"
             Format="urn:oasis:names:tc:SAML:1.1:nameid-format:emailAddress" >validuser@email.com</saml:NameID>
   <samlp:SessionIndex>id-PIa8uv0MUJARF4baBPex90LLc2TOP3nQi9k7sHNH</samlp:SessionIndex>
</samlp:LogoutRequest>
```

Successful request
```
<samlp:LogoutRequest xmlns:samlp="urn:oasis:names:tc:SAML:2.0:protocol"
                     xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion"
                     ID="_3172ae146a9c222241a190b343f69851edd867c2d5"
                     Version="2.0"
                     IssueInstant="2023-08-30T00:20:47Z"
                     Destination="https://IdPdomain"
                     >
    <saml:Issuer>https://moodledomain/auth/saml2/sp/metadata.php</saml:Issuer>
    <saml:NameID Format="urn:oasis:names:tc:SAML:1.1:nameid-format:emailAddress">validuser@email.com</saml:NameID>
    <samlp:SessionIndex>id-t4NmuHETnY4gQHS5UQwYXy2kv1HmXkbxdYBBT1tZ</samlp:SessionIndex>
</samlp:LogoutRequest>
```

Hoping to make this a setting in the SAML app, to expose setting/unsetting these attributes so the user can successfully logout.

Using the docs: https://simplesamlphp.org/docs/stable/simplesamlphp-reference-sp-remote.html - tried setting multiple config settings in https://github.com/catalyst/moodle-auth_saml2/blob/MOODLE_39_STABLE/config/config.php or https://github.com/catalyst/moodle-auth_saml2/blob/MOODLE_39_STABLE/config/authsources.php without any luck.

For testing, I had to edit the following lib files directly (might have been one more)
https://github.com/catalyst/moodle-auth_saml2/blob/MOODLE_39_STABLE/.extlib/simplesamlphp/modules/saml/lib/Auth/Source/SP.php#L601
https://github.com/catalyst/moodle-auth_saml2/blob/MOODLE_39_STABLE/.extlib/simplesamlphp/modules/saml/lib/IdP/SAML2.php#L1270
https://github.com/catalyst/moodle-auth_saml2/blob/MOODLE_39_STABLE/.extlib/simplesamlphp/modules/core/lib/Auth/Process/TargetedID.php#L151

### PR DESCRIPTION
**Pull request by brendanheywood**
Closes #769

### CODE REVIEW

File: .extlib/simplesamlphp/lib/SimpleSAML/Utils/Config/Metadata.php [hunk 1]
```diff
@@ -330,10 +330,10 @@ public static function parseNameIdPolicy($nameIdPolicy)
                 'Format'      => $nameIdPolicy_cf->getString('Format', Constants::NAMEID_TRANSIENT),
                 'AllowCreate' => $nameIdPolicy_cf->getBoolean('AllowCreate', true),
             ];
-            $spNameQualifier = $nameIdPolicy_cf->getString('SPNameQualifier', false);
-            if ($spNameQualifier !== false) {
-                $policy['SPNameQualifier'] = $spNameQualifier;
-            }
+            //$spNameQualifier = $nameIdPolicy_cf->getString('SPNameQualifier', false);
```
Comment (brendanheywood) [inline .extlib/simplesamlphp/lib/SimpleSAML/Utils/Config/Metadata.php:333]: so reverse engineering this, I think this line:

https://github.com/catalyst/moodle-auth_saml2/blob/MOODLE_39_STABLE/config/authsources.php#L65C5-L65C56

needs to change into something like:

```
'NameIDPolicy' => [
  'Format' => $saml2auth->config->nameidpolicy,
  'AllowCreate' => ???,
  'SPNameQualifier' => $newsettinghere
],
```

Comment (jwalits) [reply .extlib/simplesamlphp/lib/SimpleSAML/Utils/Config/Metadata.php:333]: I did try that, but didn't work - will give it another shot in the morning..

**Comment by jwalits** [comment]
Tried 3 different scenarios as discussed

a) Update authsources to convert NameIdPolicy to an array


This made no difference to the logout request - SPNameQualifier was still getting passed and couldn't log out of Moodle.

b) Change nameid to transient
nameidtransient

This caused the following error when trying to login from Moodle
nameidtransient_error

c) Set "Expose NameID as attribute" setting to No. Redirects to IdP correctly, but after successful login at IdP level, it fails to login to Moodle



However, when "Logout" is clicked above, the SPNameQualifier attribute is not sent and logout from IdP is successful

**Comment by jwalits** [comment]
Probably no need for this PR now, as we've found a way to avoid doing this by choosing "No" for "Expose Nameid as attribute" option and passing email address as a regular field in the saml response from the IdP.

**Pull request closed without merging**